Feature: Link from code to GitHub via context menus

shana opened the pull request:
Fixes #106 

This needs:

- [x] PR #187 - New menu handling system in master
- [x] PR #220 - Status bar notification service
- [x] PR #221 - Octicons update
- [x] Notification when link is copied on the status bar
- [ ] Code review

shana commented:
This is how it looks like currently.

@donokuda @Haacked Should we keep these context menu entries like this (considering we have potentially more context menus to add), or should we have a GitHub context menu entry with these as submenus?

![copylink](https://cloud.githubusercontent.com/assets/310137/12900383/5357ca46-ceb6-11e5-9c01-2aaa6f0eaa79.gif)

shana commented:
@Haacked This is ready for :eyes:. A few things in here are part of PR #220 so might want to deal with that one first.

Haacked commented:
People gonna be upset if we add too much to the context menu. If there isn't already a grouping that makes sense, we should group our commands in a top level "GitHub" menu.

donokuda commented:
> People gonna be upset if we add too much to the context menu. If there isn't already a grouping that makes sense, we should group our commands in a top level "GitHub" menu.

:+1: I can see us having to do this anyway as we integrate more features (such as creating a gist.)

Haacked commented on src/GitHub.Exports/Extensions/SimpleRepositoryModelExtensions.cs:
```diff
@@ -30,5 +35,58 @@ public static bool MightContainSolution(this ISimpleRepositoryModel repository)
+        public static string CurrentSha(this ISimpleRepositoryModel repository)
+        {
+            var repo = GitService.GitServiceHelper.GetRepo(repository.LocalPath);
+            return repo.Commits.FirstOrDefault()?.Sha;
```
Could `repo` be null here? Should this be `repo?.Commits.FirstOrDefault()?.Sha`?

Haacked commented on src/GitHub.Exports/Extensions/SimpleRepositoryModelExtensions.cs:
```diff
@@ -30,5 +35,58 @@ public static bool MightContainSolution(this ISimpleRepositoryModel repository)
+        }
+
+        public static Uri GenerateUrl(this ISimpleRepositoryModel repository, string path = null,
+            int startLine = -1, int endLine = -1)
```
I'd like to avoid optional arguments in public APIs because of potential ambiguities and versioning issues: http://haacked.com/archive/2010/08/10/versioning-issues-with-optional-arguments.aspx/

Haacked commented on src/GitHub.Exports/Extensions/SimpleRepositoryModelExtensions.cs:
```diff
@@ -30,5 +35,58 @@ public static bool MightContainSolution(this ISimpleRepositoryModel repository)
+
+            var commitSha = repository.CurrentSha();
+            var uri = GenerateUrl(cloneUrl.ToRepositoryUrl().AbsoluteUri, commitSha, path, startLine, endLine);
+            return new UriString(uri).ToUri();
```
On line 66 you generate a URL using `cloneUrl.ToRepositoryUrl().AbsoluteUri` so you know this is going to be an `http` or `https` URL (that's what `ToRepositoryUrl` guarantees. So you _could_ simplify the previous two lines to...

```csharp
var uri = GenerateUrl(cloneUrl.ToRepositoryUrl().AbsoluteUri, commitSha, path, startLine, endLine);
return new Uri(uri);
```

Alternatively, because we have an implicit conversion, you could also simplify it to...

```csharp
UriString uri = GenerateUrl(cloneUrl.ToRepositoryUrl().AbsoluteUri, commitSha, path, startLine, endLine);
return new uri.ToUri();
```

I leave it to you. There's nothing explicitly wrong with what you have here. :smile:

Haacked commented on src/GitHub.Exports/Extensions/SimpleRepositoryModelExtensions.cs:
```diff
@@ -30,5 +35,58 @@ public static bool MightContainSolution(this ISimpleRepositoryModel repository)
+
+        static string GenerateUrl(string basePath, string sha, string path, int startLine = -1, int endLine = -1)
+        {
+            var sb = new StringBuilder(basePath);
```
For small number of string concatenations, I usually don't choose to use `StringBuilder` because the overhead of it isn't paid off until you get to a certain number of concatenations.

Given this is not performance critical code (it's not called often and not in a tight loop) I think it probably doesn't matter either way.

Jon Skeet has a great blog post about the subject. http://jonskeet.uk/csharp/stringbuilder.html

Haacked commented on src/GitHub.VisualStudio/Helpers/ActiveDocumentSnapshot.cs:
```diff
@@ -0,0 +1,38 @@
+{
+    [Export(typeof(IActiveDocumentSnapshot))]
+    [PartCreationPolicy(CreationPolicy.NonShared)]
+    class ActiveDocumentSnapshot : IActiveDocumentSnapshot
```
Can we export private classes?

Haacked replied:
*bump* I'm surprised this works. Maybe I shouldn't be?

shana replied:
It works via reflection, so it doesn't actually care about the visibility of the implementation. But it should be public, yes, for consistency.

Haacked commented on src/GitHub.VisualStudio/Menus/CopyLink.cs:
```diff
@@ -0,0 +1,41 @@
+            var link = GenerateLink();
+            if (link == null)
+                return;
+            Clipboard.SetText(link.AbsoluteUri);
```
The documentation suggests `Clipboard.SetText` can only throw an `ArgumentNullException`. We've learned the hard way that's not true. For example, the [.NET source code](http://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/Clipboard.cs,139) shows it could throw a `SecurityException`.

Also, this interacts with COM and what happens in COM land, doesn't  stay in COM land but craps potential exceptions all over your best laid plans.

So we should probably wrap this in a `try/catch` so it doesn't crash the extension if adding to the Clipboard fails. In GitHub Desktop, we retry a couple times in case it's some weird intermittent failure. We don't have good data to know if it's possible that a retry would succeed though. So I'd just wrap it in a try/catch, write out a message, and the user can try again.

shana replied:
The `System.Windows` Clipboard behaves differently from the `System.Windows.Forms` one, but yeah, it can still throw completely random exceptions whenever it feels like it. I'll follow your suggestion and print out a message if it fails.

Haacked commented:
:fire_engine:

shana commented:
I've moved things to a submenu and cleaned up. The icons look like crap though, even though they're xaml icons. Feels like the xaml versions of the octicons that we have have strokes that are too wide?
![screen shot 2016-02-09 at 2 04 53 pm](https://cloud.githubusercontent.com/assets/310137/12916944/8073d094-cf36-11e5-879d-97420599c504.png)
![screen shot 2016-02-09 at 2 05 09 pm](https://cloud.githubusercontent.com/assets/310137/12916948/86c5ab16-cf36-11e5-808f-94c554478980.png)

Haacked commented:
I think we need an :octocat: icon for the top level GitHub menu item. What do you think @donokuda and @shana?

shana commented:
@haacked Would love to, but unfortunately it seems VS does not support icons on menu entries that lead to submenus :cry:

Haacked commented:
> @haacked Would love to, but unfortunately it seems VS does not support icons on menu entries that lead to submenus

:cry: /cc @acangialosi feature request!

Haacked commented:
@shana I just have the one question about the private class that's exported.

shana commented:
@Haacked I've gone and :hammer: all the things you mentioned. I still need to fix the icons so they look half decent, so don't merge this quite yet.

Haacked commented:
_Slowly takes finger off the big green button_

shana commented:
![](https://media.giphy.com/media/hDSy8w6rGHeTe/giphy.gif)

niik commented:
@shana A+

shana commented:
@Haacked 

Fixed the icon sizes. Also fixed the colors to match the colors of other context menus in VS. In the process, added a little extra code to support theming, so now we have three separate theme xaml files that define colors for the three themes VS comes with - Dark, Light and Blue.

Any xaml file that includes a SharedDictionaryManager resource dictionary with the uri "Theme.xaml" will get a themed dictionary included instead that tracks the VS theme and gets switched at runtime whenever the VS theme changes. If VS is running an unknown theme, the code detects whether it's a dark or light theme (based on brightness) and loads our dark or light theme file to match.

# Test run

green for dark theme, blue for blue theme, red for light theme
![runtime-theme-change](https://cloud.githubusercontent.com/assets/310137/13190737/54958f72-d75f-11e5-859f-dd9c22a83cac.gif)

# Dark
![screen shot 2016-02-19 at 11 09 46 pm](https://cloud.githubusercontent.com/assets/310137/13190891/5dbe0862-d760-11e5-9240-50f1fd2c9ab0.png)

# Light
![screen shot 2016-02-19 at 11 11 18 pm](https://cloud.githubusercontent.com/assets/310137/13190895/6480af24-d760-11e5-9827-0288d2ad638a.png)

# Blue
![screen shot 2016-02-19 at 11 11 04 pm](https://cloud.githubusercontent.com/assets/310137/13190900/69fced8c-d760-11e5-829c-b020da53ad13.png)

donokuda commented:
> Any xaml file that includes a SharedDictionaryManager resource dictionary with the uri "Theme.xaml" will get a themed dictionary included instead that tracks the VS theme and gets switched at runtime whenever the VS theme changes. If VS is running an unknown theme, the code detects whether it's a dark or light theme (based on brightness) and loads our dark or light theme file to match.

![tumblr_nozh54mtna1s31h01o1_400](https://cloud.githubusercontent.com/assets/1174461/13191083/683fcd80-d705-11e5-930e-3793269c1999.gif)

> green for dark theme, blue for blue theme, red for light theme

Are these colors just for demonstrating how theming works?

shana commented:
> Are these colors just for demonstrating how theming works?

@donokuda Yeah, I couldn't actually tell the difference. I'm actually using the same color for those icons on all three themes, and VS is apparently inverting the color on the dark theme when rendering the icons (got me so confused initially when I set it to white and it showed up black and vice-versa...), so I set them to RGB to make sure it was working (and recorded proof for future self)

donokuda commented:
:+1: That's what I thought. Love this!

Haacked commented:
This feature feels like magic!

Atcold commented:
How does one activate this?
